### PR TITLE
feat(SPM-1707): paginate remediation pair fetching

### PR DESCRIPTION
**Pull request by mkholjuraev**
This will enable paginating the request to fetch remediation pairs on Systems and Advisories pages.

To test:
1. Visit the advisories page
2. bulk select a page/all rows
3. click on Remediate
4. Observe that there are multiple requests made to fetch paginated result
5. Observe that those results are being aggregated correctly
6. Observe that Remediation modal opens up as expected and you can successfully remediate
7. Please do so on the Systems page as well.

### CODE REVIEW

**Review by gkarat — APPROVED** [review]
Looks OK to me, tried locally and the requests are chunked as expected on both pages. But there is a couple of things I want you to improve, please see the comments.

---
Also one thing, rather a minor bug:
1. Go to the Advisories page and select all items in the table.
2. Then try to select only the current page (20 items by default).

image

You won't be able to do that and even after you select only the page items, the bulk select still displays that all systems are selected. You can fix it by clicking "-" in the checkbox, and then select the page. But I would expect you can also achieve this with both ways.

**Comment by mkholjuraev** [comment]
@gkarat your comment regarding the Advisor page selection does not comply with the UX guidelines. We explicitly implemented bulk selection in such a way that after all items are selected in the table, you can not select a page.

File: src/Utilities/useRemediationDataProvider.js [hunk 1]
```diff
@@ -1,3 +1,4 @@
+import { chunk } from 'lodash';
```
[the remark this answers is not in the document]
Comment (mkholjuraev) [reply src/Utilities/useRemediationDataProvider.js:1]: Nice suggestion.

File: src/Utilities/useRemediationDataProvider.js [hunk 2]
```diff
@@ -10,9 +11,23 @@ import {
...
+const PAIRS_CHUNK_SIZE = 100;
+
+export const prepareRemediationPairs = (payload = [], remediationType) => {
+    const fetchFunction = remediationType === 'systems' ? fetchViewSystemsAdvisories : fetchViewAdvisoriesSystems;
+    const payloadChunks = chunk(payload, PAIRS_CHUNK_SIZE);
+
+    const fetchPromises = [];
+    for (let i = 0; i < payloadChunks.length; i++) {
+        fetchPromises.push(fetchFunction({ [remediationType]: payloadChunks[i] }));
+    }
```
[the remark this answers is not in the document]
Comment (mkholjuraev) [reply src/Utilities/useRemediationDataProvider.js:23]: I find this to be easier to read :)

File: src/Utilities/useRemediationDataProvider.test.js [hunk 3]
```diff
@@ -46,3 +50,58 @@ describe('useRemediationDataProvider', () => {
...
+    }
+
+    it('Should create correct number of api requests with chunk size of 100', async () => {
+        const advisoryRemediationProvider = useRemediationDataProvider(manySystems, setRemediationLoading, 'systems');
+        await act(() => advisoryRemediationProvider());
+
+        expect(fetchViewSystemsAdvisories).toHaveBeenCalledTimes(2);
+    });
+
+    it('Should create only one api requests with chunk size of 100', async () => {
```
[the remark this answers is not in the document]
Comment (mkholjuraev) [reply src/Utilities/useRemediationDataProvider.test.js:67]: Nope actually, this is correct as there is only one selected item and the API chunk size is 100. Thus only one API call is enough. The number of API calls is the total number across tests, we need to consider the previous 2 calls as well.

**Comment by gkarat** [comment]
ok, then PR is clean, LGTM

**Comment by mkholjuraev** [comment]
@gkarat thank you for the review!

**Pull request merged**